Filipchik_Kirill

filipchikkirill opened the pull request.

filipchikkirill committed:
test from kirill

ArtemNikolaev1 committed:
Initial commit v5

ArtemNikolaev1 committed:
Gitignore update

ArtemNikolaev1 committed:
Moving and gitignore update

filipchikkirill committed:
add homework_lesson_2

filipchikkirill committed:
add homework_lesson_2

filipchikkirill committed:
homework_lesson_2
add StringBuilder
remove case 5-9,0

filipchikkirill committed:
rename Homework_lesson_2 -> homework
rename MyApp.java -> Main.java

filipchikkirill committed:
refactor README.md

filipchikkirill committed:
Delete src/main/java/lesson_2 directory

NikolaevArtem commented:
Please merge with master

NikolaevArtem commented on .gitignore:
```diff
@@ -1,4 +1,37 @@
 /build/
-/build/classes/java/main/
\ No newline at end of file
+/build/classes/java/main/
```
Please merge with master

NikolaevArtem commented on src/main/java/homework_1/Main.java:
```diff
@@ -0,0 +1,35 @@
+        for (String arg :
+                args) {
+            if (arg.equalsIgnoreCase("ошибка")) {
+                System.err.println("Тревога!");
```
Doesn't work as expected

NikolaevArtem commented on src/main/java/homework_1/Main.java:
```diff
@@ -0,0 +1,35 @@
+
+public class Main {
+    public static void main(String... args) {
+        if (args.length == 0) return;
```
Not needed

NikolaevArtem commented on src/main/java/homework_1/Main.java:
```diff
@@ -0,0 +1,35 @@
+                System.err.println("Тревога!");
+                return;
+            } else {
+                strBuilder
```
Looks complicated, better simplify or move to a method

filipchikkirill committed:
fix bugs
moved the creation of result String -> new method createStringResult
removed check for missing args

filipchikkirill committed:
fix bugs
add color -> instead -> System.err
the last commit deleted the main, returned it.

filipchikkirill committed:
Merge remote-tracking branch 'origin/master' into feature/KirillFilipchik

# Conflicts:
#	README.md
#	src/main/java/homework_1/Main.java

filipchikkirill committed:
fix bugs

NikolaevArtem commented on src/main/java/homework_1/Main.java:
```diff
@@ -1,9 +1,39 @@
+    public static void main(String... args) {
+        for (String arg :
+                args) {
+            if (arg. equals("ошибка")) {
```
Opt: please fix style according code conventions

filipchikkirill committed:
fix style
conditions in one line

filipchikkirill committed:
homework 2 start
add ResultPrinter.java

filipchikkirill committed:
homework 2 PyramidPrinter
add PyramidPrinter.java

filipchikkirill committed:
homework 2 change README.md

filipchikkirill committed:
homework 2 add TrafficLight.
rework PyramidPrinter -> OldPyramidPrinter
all launch in Main.java

filipchikkirill committed:
homework 2 add RandomCharsTable.
change README.md

filipchikkirill committed:
homework 2 ready for check.
add abstract classes App and AppWithIntInput
add interface Executable
removed duplicate code
I cant start new branch, so del all old.

filipchikkirill committed:
homework 2 ready for check.
remade enum Color and as a consequence class TrafficLight

filipchikkirill committed:
homework 2 ready for check.
fixes.
add check value in RandomCharsTable(incomplete as correct behavior is not specified)

filipchikkirill committed:
homework 2 refactoring the project structure

filipchikkirill committed:
homework 2 RandomCharsTable refactor

filipchikkirill committed:
homework 2 refactor TrafficLight

filipchikkirill committed:
homework 2 refactor utils
App.java - change err msg
AppWithIntInput.java -> AppWithOneIntInput.java

filipchikkirill committed:
modify README.md

filipchikkirill committed:
homework_3

filipchikkirill committed:
homework_2 small refactor

filipchikkirill committed:
homework_2 small refactor

filipchikkirill committed:
homework_2 I don't know what result is needed, so now output the raw result

filipchikkirill committed:
homework_2 method start() -> run()

filipchikkirill committed:
Merge remote-tracking branch 'origin/master' into feature/KirillFilipchik

filipchikkirill committed:
homework_3 add PyramidPrinterTest

filipchikkirill committed:
homework_3 add TrafficLightTest

filipchikkirill committed:
homework_3 add RandomCharsTableTest not full

filipchikkirill committed:
homework_2 fix RandomCharsTable
fix check strategy
add method setDebug which sets the seed(1L) for the randomizer

filipchikkirill committed:
homework_3 ready for check

filipchikkirill committed:
homework_3 ready for check
remove test RandomCharsTableTest.InputValidArgumentsNotSameResults() (can't to do it)

filipchikkirill committed:
homework_3 ready for check
README.md update

filipchikkirill committed:
homework_3 ready for check
add some tests.

filipchikkirill committed:
homework_3 ready for check
fix tests in RandomCharsTableTest
InputCheckStrategyOdd and InputCheckStrategyEven -> InputRandomCheckStrategy

filipchikkirill committed:
homework_3 ready for check
add @RepeatedTest(10)

filipchikkirill committed:
homework_3 ready for check
add Tests for HW_1

NikolaevArtem commented on src/main/java/homework_2/pyramid_printer/Main.java:
```diff
@@ -0,0 +1,11 @@
+public class Main {
+
+    public static void main(String[] args) {
+        Executable pyramidPrinter = new PyramidPrinter();
+        pyramidPrinter.run();
```
Opt: could be simplified to 
```suggestion
        new PyramidPrinter().run();
```

NikolaevArtem commented on src/main/java/homework_2/random_chars_table/RandomCharsTable.java:
```diff
@@ -0,0 +1,75 @@
+
+import java.util.Random;
+
+public class RandomCharsTable extends App {
```
Somehow, without letters it removes the line
![image](https://user-images.githubusercontent.com/58872509/126624149-ae122cfc-2426-4a94-bef8-1eed4c99c273.png)

NikolaevArtem replied:
Good job! All tests are passed, works as expected. For style, it's good that you tried most of the things we learn, but in prod, I don't recommend doing it, as simplicity is usually the main goal

NikolaevArtem replied:
Opt: make homework simpler to undertand where possible

filipchikkirill replied:
If there are no letters suitable for the strategy, then the string - "* Strategy * letters" - should not be displayed? Did I understand correctly?
Of course, I added inheritance for the sake of inheritance. In real tasks, this is not necessary, I understand that.

NikolaevArtem commented on src/test/java/homework_2/random_chars_table/RandomCharsTableTest.java:
```diff
@@ -0,0 +1,133 @@
+    // дальше тест уродлив, но как иначе распарсить эти строки и сравнить все я не знаю
+    // за то покрывает все варианты.
+    @RepeatedTest(10)
+    void InputRandomCheckStrategy() {
```
It seems it does the trick, but possible bugs could be present. If possible, simplify :)

NikolaevArtem commented on src/test/java/homework_2/random_chars_table/RandomCharsTableTest.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    /*
+        for this test use new RandomCharsTable().setDebug(true).run();
```
If think, once you'll understand how to check random letters, you'll not need debug mode. Good idea anyway

NikolaevArtem commented on src/main/java/homework_3/ImmutableClass.java:
```diff
@@ -0,0 +1,46 @@
+No setters
+ */
+public final class ImmutableClass {
+    private final String aString;
+    private final int anInt;
+    private final List<String> aList;
```
Opt: Variables usually are not named this way

NikolaevArtem commented on src/main/java/homework_3/ImmutableClass.java:
```diff
@@ -0,0 +1,46 @@
+Deep Copy of objects should be performed in the getter methods
+No setters
+ */
+public final class ImmutableClass {
```
Good

filipchikkirill committed:
homework_2 fix Main.java

filipchikkirill committed:
homework_2 fix Main.java

filipchikkirill committed:
homework_4 start
add Singleton and SingletonTest

filipchikkirill committed:
homework_4 change README.md

filipchikkirill committed:
homework_4 add annotation

filipchikkirill committed:
homework_4 add annotation

filipchikkirill committed:
homework_4 add CustomFileReader
add resource/custom_file_reader/file

filipchikkirill committed:
homework_4 add fix README.md

filipchikkirill committed:
homework_4 add fix

filipchikkirill committed:
homework_5 add

tmandreeva1 commented on src/test/java/homework_4/custom_file_reader/CustomFileReaderTest.java:
```diff
@@ -0,0 +1,41 @@
+    void TestRun1() throws IOException {
+        run1();
+        assertFalse(getOutput().contains("."));
+        assertFalse(getOutput().contains(","));
```
Probably, what should you really test, is edge cases: what happens if file is empty? Or doesn't contain any dots or commas? Also, since you have setter for a file path in your CustomFileReader - what if we set a path of a non-existent file? 
And all of this test cases can be applied to all of the reading methods.

filipchikkirill replied:
"реализовать приложение, которое считывает данные из файла, и печатает в консоль, но уже без запятых и точек."
If there are no commas or dots in the file, or the file is empty, then these tests will pass perfectly.
The output should not contain dots and commas - this is the condition of the task. This is verified by tests.

"Файл - тот, что вы даете. Другого файла не ожидается. Все проще) "
Thanks for the comment, removed the setter.

filipchikkirill replied:
Add test - "removed only commas and dots" is a good idea

tmandreeva1 replied:
_> If there are no commas or dots in the file, or the file is empty, then these tests will pass perfectly._ 

But what if you change your run method implementation in future? Test may fail in one situation and pass in others and that is not what is expected from tests. I mean:
1. The purpose of tests is to be able to add new code, or remove, or refactor, or whatever with being assure that you're not impacting the code you weren't plan to impact, that other app features work correctly. 
2. Tests need to be atomic, which means that one test check one test case and if it fails, it's crystal clear what is wrong. If your current tests fail, you certainly will understand which method is not working as expected (run1 or run2 etc), but you will need to test your app manually to get the concrete cause of test fail. Which will cost you some time. (test passed with ok and ordinary and expected file and test passed with empty file - these are two different test cases, because they have different data given).

I know, that hw4 wasn't about testing, nor about changing app in future (which, by the way, you never know), but your answer made me feel, that you're not properly understand testing, which might be important sometime in your career. You could search for some reliable sources about testing, if you want, it's quite good to know.
Also, whatever the task is, testing edge cases is must-have, in my opinion. 

Whatever, it's just the thoughts:)

tmandreeva1 replied:
> "Файл - тот, что вы даете. Другого файла не ожидается. Все проще) " 

File - file name - could be the same, contents might differ

filipchikkirill committed:
homework_4 fix
removed setter and getter for file_path

filipchikkirill committed:
homework_4 add test for CustomAnnotation

filipchikkirill committed:
homework_5 add

filipchikkirill committed:
homework_5 README.md change

filipchikkirill committed:
homework_6
add map_problems_generator
README.md change
fix hw5 package

filipchikkirill committed:
homework_6
README.md fix

ArtemNikolaev1 commented on src/main/java/homework_4/custom_file_reader/Main.java:
```diff
@@ -0,0 +1,19 @@
+
+public class Main {
+    public static void main(String[] args) {
+        try {
```
1. Better have one try-with-resources per one stream. So it's better to have 4 try/catch in run() methods.
2. You don't close streams

filipchikkirill replied:
Done.

filipchikkirill committed:
homework_7
add
README.md fix
add SuppressWarnings in Main hw_6

filipchikkirill committed:
homework_4
fix

filipchikkirill committed:
course_project first iteration

filipchikkirill committed:
course_project fix 1

filipchikkirill committed:
course_project fix 2

filipchikkirill committed:
course_project fix 3

filipchikkirill committed:
course_project fix 4

filipchikkirill committed:
course_project fix 5

filipchikkirill committed:
course_project fix 6

filipchikkirill committed:
course_project add Tests (not all)

NikolaevArtem commented on src/main/java/homework_3/ImmutableClass.java:
```diff
@@ -0,0 +1,46 @@
+    public ImmutableClass(List<String> aList) {
+        this.aString = "default";
+        this.anInt = 0;
+        this.aList = aList;
```
Actually, I see now your class is mutable. A parameterized constructor should initialize all the fields performing a deep copy

filipchikkirill replied:
done here

NikolaevArtem commented on src/main/java/homework_3/ImmutableClass.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public ImmutableClass getNew(List<String> aList) {
+        return new ImmutableClass(aString, anInt, aList);
```
Same here

filipchikkirill replied:
I don't understand what the problem is, I redid it just in case...

NikolaevArtem commented on src/main/java/homework_4/custom_file_reader/Main.java:
```diff
@@ -0,0 +1,10 @@
+
+public class Main {
+    public static void main(String[] args) {
+        CustomFileReader.run();
```
Looks strange 
![image](https://user-images.githubusercontent.com/58872509/134134035-989d1cbe-2720-4d27-9696-12d2cb228c23.png)

filipchikkirill replied:
How do I get this error? 
remind, i don't use windows

NikolaevArtem commented on src/main/java/homework_4/custom_file_reader/CustomFileReader.java:
```diff
@@ -0,0 +1,66 @@
+    public static void run2() throws IOException {
+        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(getFileURL().openStream()));
+        bufferedReader.lines().forEach(FUNCTION);
+        bufferedReader.close();
```
Now you close streams, but not where you should. Better use try-with-resources, you'll avoid this mistake

NikolaevArtem replied:
In case you get an exception on 48th line, bufferedReader will not be closed

filipchikkirill replied:
done. Add try with res

NikolaevArtem commented on src/main/java/homework_4/singleton/Singleton.java:
```diff
@@ -0,0 +1,13 @@
+    private Singleton() {}
+
+    private static class Holder {
+        static final Singleton INSTANCE = new Singleton();
```
Opt: why not just have static variable?

filipchikkirill replied:
Lazy init.

filipchikkirill replied:
The JVM defers initializing the InstanceHolder class until it is actually used, and because the Singleton is initialized with a static initializer, no additional synchronization is needed. The first call to getInstance by any thread causes Holder to be loaded and initialized, at which time the initialization of the Singleton happens through the static initializer.

NikolaevArtem replied:
Agree, done

NikolaevArtem commented on src/main/java/homework_4/сustom_annotation/SimpleClassWithNameField.java:
```diff
@@ -0,0 +1,25 @@
+        try {
+            constructor = this.getClass().getConstructor(String.class);
+            CustomAnnotation annotation = constructor.getAnnotation(CustomAnnotation.class);
+            if (annotation != null) this.name = annotation.value();
+            if (!name.isEmpty()) this.name = name;
```
Opt: 
1) let's not use if without brackets, this approach is prone to errors
2) I think, the idea you try to implement would look better if 'if-else' or ternary operator will be used

filipchikkirill replied:
add brackets

NikolaevArtem commented on src/main/java/homework_4/сustom_annotation/CustomAnnotation.java:
```diff
@@ -0,0 +1,13 @@
+
+@Retention(RUNTIME)
+@Target(CONSTRUCTOR)
+public @interface CustomAnnotation {
```
Good!

NikolaevArtem commented on src/main/java/homework_5/AppHW_5.java:
```diff
@@ -0,0 +1,54 @@
+
+import java.util.NoSuchElementException;
+
+public class AppHW_5 {
```
Opt: in general idea to separate user input class is a good one, but we should not forget that abstractions are made when they simplify reading and fixing the code, not make it less readable

NikolaevArtem commented on src/main/java/homework_5/custom_regex_matcher/CustomRegexMatcher.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    public String run() {
+        readData();
+        String result = String.valueOf(check());
+        System.out.println(result);
+        closeReader();
+        return result;
```
Opt: as example here it's pretty unclear what is happening. Void methods usually are not recommended too

NikolaevArtem commented on src/main/java/homework_6/map_problems_generator/problem_impl/GettingProblem.java:
```diff
@@ -0,0 +1,15 @@
+
+    @Override
+    public int hashCode() {
+        return new Random().nextInt();
```
I reread the task description, and I agree it's unclear there. But the task I had in mind is when you put mutable key, change it's field, and try to get value -> you can't, because hash is now different, even though hash was written according contracts.

filipchikkirill replied:
add this case(new class add)

NikolaevArtem commented on src/main/java/homework_6/map_problems_generator/problem_impl/Wrapper.java:
```diff
@@ -0,0 +1,46 @@
+
+public class Wrapper <T> implements Problem<T> {
+    private final Problem<T> innerProblem;
+    private final List<String> event = new ArrayList<>();
```
Good implementation for generating events!

NikolaevArtem commented on src/main/java/homework_7/Main.java:
```diff
@@ -0,0 +1,16 @@
+
+    public static void main(String[] args) {
+        Kitten kitten = new Kitten("Murzik", 1);
+        KittenToCatFunction function = (k) -> new Cat(k.getName(), k.getAge() + 4);
```
Opt: you could add some more complicated logic to lambda, as example - converting one type to another (like, favourite toy to favourite color or smth like this)

NikolaevArtem commented on src/main/java/course_project/sea_battle/SeaBattleApp.java:
```diff
@@ -0,0 +1,9 @@
+
+import course_project.sea_battle.controller.impl.GameControllerImpl;
+
+public class SeaBattleApp {
```
Opt: The field is quite narrow, i think it would look better with more space between cells by width
![image](https://user-images.githubusercontent.com/58872509/134136786-52412ebe-bf2e-4eaa-80b7-50242fad8030.png)

NikolaevArtem replied:
The game seems not working. I see you implemented PvP, but I play only for 1 Player. I don't see field for 2nd, don't see his name anywhere. Don't know how to shoot for 2nd. 
And please introduce automatic ship placement, testing will be much easier

NikolaevArtem replied:
Good! Nice architecture, SOLID is followed. Code is pretty readable. Good Java Core and Java libraries usage. 

To make better: interface could be refactored. Also check correct working of the game.

NikolaevArtem replied:
Also nice approach to testing, test cases are readable

NikolaevArtem replied:
Good, automatic ship placement! I found some minor issues with the interface: 
1) When miss, there's no message to press enter to finish the turn. Actually, I didn't see this message anywhere... Confusing
2) In case i hit a ship, i need to press enter 2 times to be able to shoot again
3) No error message in case I already shot this cell
![image](https://user-images.githubusercontent.com/58872509/134474624-47363b8a-6b01-4739-a488-2402b98bb992.png)
4) No message in case of destroyed ship 
![image](https://user-images.githubusercontent.com/58872509/134474815-1c7607f9-a49a-43ed-89f7-778bd4e5156f.png)
5) And please make field representation wider :) 

The app now works correctly, I didn't find any bugs. And from the code perspective, the app looks great! 

To make better: improve interface:) 
The homework is approved

NikolaevArtem commented on src/test/java/homework_4/singleton/SingletonTest.java:
```diff
@@ -0,0 +1,16 @@
+        Singleton singleton1 = Singleton.getInstance();
+        Singleton singleton2 = Singleton.getInstance();
+
+        assertEquals(singleton1, singleton2);
```
Opt: they can be equal, but be different objects, right?

filipchikkirill replied:
they share a common pointer. (added this check (IDE don't like this))

filipchikkirill committed:
Homework3 fix immutableclass

NikolaevArtem reviewed (changes requested):
Nice! The code needs small fixes, but except that - good job! The architecture is nice, and I see you exercised with inheritance/generics, which is cool. But remember that it's nice for learning purposes, but production code, at best, should be readable and simple first. Usually, when there's a choice between SOLID or readability/simplicity, I choose second. IMHO

filipchikkirill committed:
Homework3 fix immutableclass

filipchikkirill committed:
Homework 6 add mutable

filipchikkirill committed:
Homework 3 add mutable

filipchikkirill committed:
Homework fix SingletonTest

filipchikkirill committed:
course_project fix and add random fill field

filipchikkirill committed:
course_project fix

filipchikkirill committed:
course_project fix

filipchikkirill committed:
course_project test

filipchikkirill committed:
course_project test

filipchikkirill committed:
course_project test

filipchikkirill committed:
course_project test

filipchikkirill committed:
hw_3 fix

NikolaevArtem commented on src/main/java/homework_4/custom_file_reader/CustomFileReader.java:
```diff
@@ -0,0 +1,66 @@
+    private static final Consumer<String> FUNCTION = x -> System.out.println(x.replaceAll("[,.]", ""));
+
+    public static void run() {
+        try {
+            run1();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        try {
+            run2();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        try {
+            run3();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        try {
+            run4();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Opt: these try-catch now are either not needed or not in right place...

NikolaevArtem reviewed (approved):
Approved!
